File: src/pkg/resolve.ts
import { readFileSync, existsSync } from "fs";
import { join } from "path";
import { Lockfile } from "./lock";
import { fetchDep } from "./fetch";
import { Engine } from "../diag/engine";
import { Code } from "../diag/code";

/**
 * Dependency Resolution
 * 
 * DESIGN DECISION: Packages declare artifacts strictly via `project.json`.
 * We deliberately DO NOT support arbitrary build hooks (e.g., `preinstall`, `postbuild`)
 * in dependencies. This ensures reproducible builds, prevents supply-chain attacks
 * via malicious build scripts, and keeps the dependency graph purely declarative.
 */

export interface Manifest {
  name: string;
  version: string;
  dependencies?: Record<string, string>;
}

export interface ResolvedDep {
  name: string;
  source: string;
  hash: string;
  path: string;
  requiredBy: string;
}

export class ResolveError extends Error {
  constructor(msg: string) { super(msg); }
}

export function resolveDeps(root: string, manifest: Manifest, lock: Lockfile | null, diag: Engine): ResolvedDep[] {
  const resolved = new Map<string, ResolvedDep>();
  const queue: { name: string; source: string; requiredBy: string }[] = [];

  if (manifest.dependencies) {
    for (const [name, source] of Object.entries(manifest.dependencies)) {
      queue.push({ name, source, requiredBy: "root" });
    }
  }

  while (queue.length > 0) {
    const { name, source, requiredBy } = queue.shift()!;
    
    if (resolved.has(name)) {
      const existing = resolved.get(name)!;
      if (existing.source !== source) {
        diag.emit(Code.PkgConflict, { start: 0, end: 0, line: 0, col: 0 }, 
          `Conflict: '${name}' required as '${source}' by '${requiredBy}', ` +
          `but already resolved as '${existing.source}' by '${existing.requiredBy}'`);
        throw new ResolveError("Dependency conflict");
      }
      continue;
    }

    const locked = lock?.packages.find(p => p.name === name && p.source === source);
    let path: string;
    let hash: string;

    const fetched = fetchDep(name, source, root);
    path = fetched.path;
    hash = fetched.hash;

    if (locked && locked.hash !== hash) {
      diag.emit(Code.PkgHashMismatch, { start: 0, end: 0, line: 0, col: 0 }, 
        `Hash mismatch for '${name}': expected ${locked.hash}, found ${hash}. ` +
        `The cache may be tampered or corrupted.`);
      throw new ResolveError("Hash mismatch");
    }

    resolved.set(name, { name, source, hash, path, requiredBy });

    const depManifestPath = join(path, "project.json");
    if (existsSync(depManifestPath)) {
      const depManifest: Manifest = JSON.parse(readFileSync(depManifestPath, "utf8"));
      if (depManifest.dependencies) {
        for (const [depName, depSource] of Object.entries(depManifest.dependencies)) {
          queue.push({ name: depName, source: depSource, requiredBy: name });
        }
      }
    }
  }

  return Array.from(resolved.values());
}

File: src/pkg/resolve.nim
import std/[json, os, sequtils, strutils, tables]
import ../diag/[code, engine, span]
import ./[fetch, lock]

type
  Manifest* = object
    name*: string
    version*: string
    dependencies*: Table[string, string]
  ResolvedDep* = object
    name*: string
    source*: string
    hash*: string
    path*: string
    requiredBy*: string
  ResolveError* = object of CatchableError

proc resolveError(message: string): ref ResolveError =
  new(result)
  result.msg = message

proc resolveDeps*(root: string; manifest: Manifest; lock: ref Lockfile; diag: Engine): seq[ResolvedDep] =
  var queue: seq[tuple[name, source, requiredBy: string]]
  for name, source in manifest.dependencies: queue.add((name: name, source: source, requiredBy: "root"))
  var resolved = initTable[string, ResolvedDep]()
  var position = 0
  while position < queue.len:
    let item = queue[position]; position.inc
    if item.name in resolved:
      let existing = resolved[item.name]
      if existing.source != item.source:
        diag.emit(PkgConflict, Span(), "Conflict: '" & item.name & "' required as '" & item.source & "' by '" & item.requiredBy & "', but already resolved as '" & existing.source & "' by '" & existing.requiredBy & "'")
        raise resolveError("Dependency conflict")
      continue
    let fetched = fetchDep(item.name, item.source, root)
    if lock != nil:
      for locked in lock.packages:
        if locked.name == item.name and locked.source == item.source and locked.hash != fetched.hash:
          diag.emit(PkgHashMismatch, Span(), "Hash mismatch for '" & item.name & "': expected " & locked.hash & ", found " & fetched.hash & ".")
          raise resolveError("Hash mismatch")
    let dependency = ResolvedDep(name: item.name, source: item.source, hash: fetched.hash, path: fetched.path, requiredBy: item.requiredBy)
    resolved[item.name] = dependency
    result.add(dependency)
    let depManifest = fetched.path / "project.json"
    if fileExists(depManifest):
      let node = parseJson(readFile(depManifest))
      if node.hasKey("dependencies") and node["dependencies"].kind == JObject:
        for name, source in node["dependencies"]: queue.add((name: name, source: source.getStr(), requiredBy: item.name))
